micromanager_gui/main_window.py

```diff
@@ -148,92 +156,105 @@ def add_frame_explorer(self, name, array):
-
-                #timepoints
-                if event_index_t > 0 and event_index_z == 0 and event_index_c == 0 and event_index_p == 0:
-                    empty_im = np.empty(((1,p_stack_length,z_stack_length,c_stack_length,) + image.shape), dtype=np.uint16)
-                    layer.data = np.concatenate((layer.data,empty_im), axis=-6)
-
-
-            if sequence.axis_order == 'tpcz' or sequence.axis_order == 'ptcz':
-                #zpositions
-                if event_index_z > 0 and event_index_c == 0 and event_index_p == 0 and event_index_t == 0:
-                    empty_im = np.empty(((1,)*len(sequence.axis_order) + image.shape), dtype=np.uint16) 
-                    layer.data = np.concatenate((layer.data,empty_im), axis=-4)
+            for a, v in enumerate(im_idx):
+                self.viewer.dims.set_point(a, v)
+
+            #save each image in the temp folder, 
+            image_name = f'{im_idx}.tif'
+            savefile = temp_folder / image_name
+
+            tifffile.tifffile.imsave(str(savefile), image)
+
+            #and delete temp folder and files when mda is done
+            if tuple(i+1 for i in im_idx) == seq.shape:
+
+                shutil.rmtree(temp_folder.parent)
                 
-                #channels
-                if event_index_c > 0 and event_index_z == 0 and event_index_p == 0 and event_index_t == 0:
-                    empty_im = np.empty(((1,)*(len(sequence.axis_order)-2) + (z_stack_length,1,) + image.shape), dtype=np.uint16)
-                    layer.data = np.concatenate((layer.data,empty_im), axis=-3)
-
-                #xypositions
-                if event_index_p > 0 and event_index_c == 0 and event_index_z == 0 and event_index_t == 0:
-                    empty_im = np.empty(((1,)*(len(sequence.axis_order)-2) + (z_stack_length,c_stack_length,) + image.shape), dtype=np.uint16)
-                    layer.data = np.concatenate((layer.data,empty_im), axis=-5)
-
-                #timepoints
-                if event_index_t > 0 and event_index_z == 0 and event_index_c == 0 and event_index_p == 0:
-                    empty_im = np.empty(((1,p_stack_length,z_stack_length,c_stack_length,) + image.shape), dtype=np.uint16)
-                    layer.data = np.concatenate((layer.data,empty_im), axis=-6)
-
-            layer.data[event_index_t,event_index_p,event_index_z,event_index_c, ...] = image
-         
-            #set which dimension to display in napari viewer -> viewer.dims.set_point(axis, index)
-            self.viewer.dims.set_point(3, event_index_c)
-            self.viewer.dims.set_point(2, event_index_z)
-            self.viewer.dims.set_point(1, event_index_p)
-            self.viewer.dims.set_point(0, event_index_t)   
+            #save layer when acquisition is finished
+            if self.mda.save_groupBox.isChecked():
 
-        except KeyError:
+                if tuple(i+1 for i in im_idx) == seq.shape:
```

Comment:
if you want this all to happen when the sequence is finished, this shouldn't go in `_on_mda_frame`... instead you should connect to the [onMDAFinished](https://github.com/tlambert03/pymmcore-remote/blob/044ff0730949725bac97a167fa63b2f0ed71c9c5/pymmcore_remote/_mmcore_plus.py#L114) signal.  (If you need that signal to provide the sequence object, you could submit a PR for that

Reply:
I managed to do it, It's super cool!!! :)

Reply:
Do you think is worth it to delete the temp folder also  onMDACanceled?

Reply:
if you look through the [events emitted by `run_mda`](https://github.com/tlambert03/pymmcore-remote/blob/044ff0730949725bac97a167fa63b2f0ed71c9c5/pymmcore_remote/_mmcore_plus.py#L72) you'll notice that if a sequence is canceled, it just breaks the for loop and goes right to the `self.emit_signal("onMDAFinished")` ... so as long as you hook up onMDAFinished to tempDir deletion, you'll be good